src/test/scala/is/hail/rvd/OrderedRVDPartitionerSuite.scala

```diff
@@ -8,37 +8,125 @@ import org.scalatest.testng.TestNGSuite
+
+    assert(partitioner.getSafePartitionLowerBound(Row(2, 7, 5)) == 0)
+    assert(partitioner.getSafePartitionUpperBound(Row(2, 7, 5)) == 1)
+
+    assert(partitioner.getSafePartitionLowerBound(Row(4, 2, 1, 2.7, "bar")) == 0)
+
+    assert(partitioner.getSafePartitionLowerBound(Row(7, 9, 7)) == 2)
+    assert(partitioner.getSafePartitionUpperBound(Row(7, 9, 7)) == 2)
+
+    assert(partitioner.getSafePartitionLowerBound(Row(11, 1, 42)) == 3)
   }
 
-  @Test def testGetPartitionPKWithLargerKeys() {
-    assert(partitioner.getPartitionPK(Row(0, 1, 3)) == 0)
-    assert(partitioner.getPartitionPK(Row(2, 7, "foo")) == 0)
-    assert(partitioner.getPartitionPK(Row(4, 2, 1, 2.7, "bar")) == 0)
-    assert(partitioner.getPartitionPK(Row(4, 3, 5)) == 1)
-    assert(partitioner.getPartitionPK(Row(7, 9, 7)) == 2)
-    assert(partitioner.getPartitionPK(Row(11, 1, 42)) == 2)
+   @Test def testGetPartitionPKWithSmallerKeys() {
+     assert(partitioner.getSafePartitionLowerBound(Row(2)) == 0)
+     assert(partitioner.getSafePartitionUpperBound(Row(2)) == 1)
+
+     assert(partitioner.getSafePartitionLowerBound(Row(4)) == 0)
+     assert(partitioner.getSafePartitionUpperBound(Row(4)) == 2)
+
+     assert(partitioner.getSafePartitionLowerBound(Row(11)) == 3)
+     assert(partitioner.getSafePartitionUpperBound(Row(11)) == 3)
+   }
+
+  @Test def testGetPartitionRange() {
+    assert(partitioner.getPartitionRange(Interval(Row(3, 4), Row(7, 11), true, true)) == Seq(0, 1, 2))
+    assert(partitioner.getPartitionRange(Interval(Row(3, 4), Row(7, 9), true, false)) == Seq(0, 1))
+    assert(partitioner.getPartitionRange(Interval(Row(-1, 7), Row(0, 9), true, false)) == Seq())
   }
 
-  // @Test def testGetPartitionPKWithSmallerKeys() {
-  //   assert(partitioner.getPartitionPK(Row(2)) == 0)
-  // }
+  @Test def testGetSafePartitionKeyRange() {
+    assert(partitioner.getSafePartitionKeyRange(Row(0, 0)).isEmpty)
+    assert(partitioner.getSafePartitionKeyRange(Row(7, 10)).isEmpty)
+    assert(partitioner.getSafePartitionKeyRange(Row(7, 11)) == Range.inclusive(2, 2))
+  }
+
+  @Test def testGenerateDisjoint() {
+    val intervals = Array(
+        Interval(Row(1, 0, 4), Row(4, 3, 2), true, false),
+        Interval(Row(4, 3, 5), Row(7, 9, 1), true, false),
+        Interval(Row(7, 11, 3), Row(10, 0, 1), true, true),
+        Interval(Row(11, 0, 2), Row(11, 0, 15), false, true),
+        Interval(Row(11, 0, 15), Row(11, 0, 20), true, false))
+
+    val p3 = OrderedRVDPartitioner.generate(Array("A", "B", "C"), kType, intervals)
+    assert(p3.satisfiesAllowedOverlap(2))
+    assert(p3.rangeBounds sameElements
+      Array(
+        Interval(Row(1, 0, 4), Row(4, 3, 2), true, false),
+        Interval(Row(4, 3, 5), Row(7, 9, 1), true, false),
+        Interval(Row(7, 11, 3), Row(10, 0, 1), true, true),
+        Interval(Row(11, 0, 2), Row(11, 0, 15), false, true),
+        Interval(Row(11, 0, 15), Row(11, 0, 20), false, false))
+    )
```

Comment:
seems like you should just write `intervals` again

Reply:
They're not the same; look at the last line. The key `[11,0,20]` can appear in both of the last intervals in `intervals`, but here we're asking for a partitioner with "partition key" all three key fields, so this isn't allowed. So `generate` puts `[11,0,20]` in the left partition.